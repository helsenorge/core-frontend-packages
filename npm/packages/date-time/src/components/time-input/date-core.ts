import moment from 'moment';

import { getHoursFromTimeString, getMinutesFromTimeString } from '@helsenorge/core-utils/date-utils';

import { TIME_SEPARATOR } from '../../constants/datetime';

export function parseDate(dateString: string): Date {
  return new Date(dateString);
}

export function buildNewDate(date: Date | undefined, timeString: string | undefined): Date | undefined {
  const hours = getHoursFromTimeString(timeString, TIME_SEPARATOR);
  const minutes = getMinutesFromTimeString(timeString, TIME_SEPARATOR);
  if (!hours || !minutes || !date) {
    return;
  }
  try {
    const momentDate = moment(date).hours(parseInt(hours, 10)).minutes(parseInt(minutes, 10));
    return momentDate.toDate();
  } catch (e) {
    return undefined;
  }
}
