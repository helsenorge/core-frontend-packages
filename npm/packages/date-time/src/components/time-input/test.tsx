import * as React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { shallow, mount, ReactWrapper } from 'enzyme';
import moment from 'moment';

import { parseDate, buildNewDate } from './date-core';

import TimeInput from '.';

describe('TimeInput', () => {
  let wrapper: ReactWrapper<{}, {}>;

  it('Renders without crashing', () => {
    shallow(<TimeInput id="1" />);
  });

  it('Renders correctly', () => {
    wrapper = mount(<TimeInput id="2" />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('Renders legend', () => {
    wrapper = mount(<TimeInput id="3" legend="Klokkeslett" />);
    expect(wrapper.render()).toMatchSnapshot();
  });

  it('Renders sublabel', () => {
    wrapper = mount(<TimeInput id="3" legend="Klokkeslett" subLabel={<span className="sublabel">{`sublLabel`}</span>} />);
    expect(wrapper.find('.sublabel').length).toEqual(1);
  });

  describe('Når man skriver inn et klokkeslett', () => {
    it('Så har inputfeltene fått riktige verdier og onTimeChange har blitt kalt', async () => {
      const handleTimeChange = jest.fn();
      render(<TimeInput onTimeChange={handleTimeChange} />);

      const hour = screen.getByLabelText('Timer');
      expect(hour).toBeVisible();
      await userEvent.type(hour, '23');

      expect((hour as HTMLInputElement).value).toBe('23');

      const minutes = screen.getByLabelText('Minutter');
      expect(minutes).toBeVisible();
      await userEvent.type(minutes, '59');

      expect((minutes as HTMLInputElement).value).toBe('59');

      expect(handleTimeChange).toHaveBeenCalledTimes(4);
      expect(handleTimeChange).toHaveBeenNthCalledWith(1, '02:');
      expect(handleTimeChange).toHaveBeenNthCalledWith(2, '23:');
      expect(handleTimeChange).toHaveBeenNthCalledWith(3, '23:05');
      expect(handleTimeChange).toHaveBeenNthCalledWith(4, '23:59');
    });
  });
});

describe('TimeInput utils', () => {
  describe('Gitt at parseDate kalles', () => {
    describe('Når den for en dato som argument', () => {
      it('så returnerer den new js date', () => {
        const d = parseDate('2017-05-18T05:28:45Z');
        expect(d).toBeDefined();
        expect(d instanceof Date).toEqual(true);
      });
    });

    describe('Når den for en dato og tidspunkt som argument', () => {
      it('så returnerer den new js date', () => {
        const d = parseDate('2017-05-18T05:28:45Z');
        const newDate = buildNewDate(d, '05:15');
        expect(newDate).not.toBeNull();

        if (newDate) {
          expect(moment(d).isSame(newDate, 'day')).toBeTruthy();
          expect(newDate.getHours()).toEqual(5);
          expect(newDate.getMinutes()).toEqual(15);
        }
      });
    });
  });
});
