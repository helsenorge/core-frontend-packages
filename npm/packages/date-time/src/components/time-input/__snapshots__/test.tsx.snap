// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput Renders correctly 1`] = `
<div
  class="mol_timeinput mol_validation"
  id="2-wrapper"
>
  <div
    aria-hidden="false"
    class="ReactCollapse--collapse"
    style="height: auto; overflow: initial;"
  >
    <div
      class="ReactCollapse--content"
    >
      <div
        aria-atomic="true"
        aria-live="assertive"
        class="mol_validation__errortext"
      />
    </div>
  </div>
  <fieldset>
    <div
      class="mol_timeinput__inputs"
    >
      <div
        class="safeInputField"
        id="2_hours-wrapper"
      >
        <span>
          <input
            aria-label="Timer"
            aria-required="false"
            autocomplete="off"
            class="hn-safe-input atom_input atom_input--medium atom_input--2"
            max="23"
            min="0"
            type="number"
            value=""
          />
        </span>
      </div>
      <span
        class="mol_timeinput__separator"
      >
        :
      </span>
      <div
        class="safeInputField"
        id="2_minutes-wrapper"
      >
        <span>
          <input
            aria-label="Minutter"
            aria-required="false"
            autocomplete="off"
            class="hn-safe-input atom_input atom_input--medium atom_input--2"
            max="59"
            min="0"
            type="number"
            value=""
          />
        </span>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`TimeInput Renders legend 1`] = `
<div
  class="mol_timeinput mol_validation"
  id="3-wrapper"
>
  <div
    aria-hidden="false"
    class="ReactCollapse--collapse"
    style="height: auto; overflow: initial;"
  >
    <div
      class="ReactCollapse--content"
    >
      <div
        aria-atomic="true"
        aria-live="assertive"
        class="mol_validation__errortext"
      />
    </div>
  </div>
  <fieldset>
    <legend>
      <span
        class="mol_timeinput__legend"
      >
        <span
          class="mol_timeinput__legend__label"
        >
          Klokkeslett
        </span>
      </span>
    </legend>
    <div
      class="mol_timeinput__inputs"
    >
      <div
        class="safeInputField"
        id="3_hours-wrapper"
      >
        <span>
          <input
            aria-label="Timer"
            aria-required="false"
            autocomplete="off"
            class="hn-safe-input atom_input atom_input--medium atom_input--2"
            max="23"
            min="0"
            type="number"
            value=""
          />
        </span>
      </div>
      <span
        class="mol_timeinput__separator"
      >
        :
      </span>
      <div
        class="safeInputField"
        id="3_minutes-wrapper"
      >
        <span>
          <input
            aria-label="Minutter"
            aria-required="false"
            autocomplete="off"
            class="hn-safe-input atom_input atom_input--medium atom_input--2"
            max="59"
            min="0"
            type="number"
            value=""
          />
        </span>
      </div>
    </div>
  </fieldset>
</div>
`;
